### ISSUE TITLE
[FEATURE] Proposal to add methods to serialize the output of documents prior to delivering to end user client

### ISSUE THREAD

**Issue by ablankenship10** [dynamoose/dynamoose#797]
## Goal
Add serialization methods to output Items/Documents in one or more formats to aid in adding, removing, or modifying information prior to output to an end user or other internal methods.

Based on this package for serializing Sequelize Documents: https://github.com/hauru/sequelize-to-json 

#### Model.addSerializer(name, options)
Configuring a new serializer that can be called by name
* name - The name of this serializer
* options - A schema defining how to manipulate a given document  

#### Model.serialize(documents: Array, nameOrOptions: String/Object/Array)
Runs an array of documents through a given serializer by name or provided schema
* documents - And array of documents belonging to this model
* nameOrOptions - One of either a name of an existing configured serializer, an array containing the names
 of fields that should be returned (equivalent to options.include), or an options object detailing more complex operations (see below)

#### Document.serialize(nameOrOptions: String/Object/Array)
Runs the given document through the named serializer or provided schema
* nameOrOptions - One of either a name of an existing configured serializer, an array containing the names
 of fields that should be returned (equivalent to options.include), or an options object detailing more complex operations (see below)

### Options

| Field | Type | Description |
| ----- | ---- | ----------- |
| include | array | Field names which should be included in the serialized output of the document |
| exclude | array | Field names which should **not** be included in the serialized output of the document |
| postSerialize | function | A function which receives the fully serialized object and the original document to make last minute changes to the output |

### Example

```javascript
const document = myModel.create({
  email: 'hello@example.com',
  username: 'bobdole',
  firstName: 'Bob',
  lastName: 'Dole',
  password: 'testing123',
  stripeCustomerId: 'cust_m1234567890',
  status: 'ACTIVE'
});

const options = {
  include: ['email', 'username', 'firstName', 'lastName'],
  exclude: ['password', 'stripeCustomerId', 'status'],
  postSerialize: (serialized, original) => {
    serialized.isActive = original.status === 'ACTIVE';
    return serialized;
  }  
}

myModel.addSerializer('mySerializer', options);

const serialized = document.serialize('mySerializer');

console.log(serialized);

/*
  Returns:

  {
    email: 'hello@example.com',
    username: 'bobdole',
    firstName: 'Bob',
    lastName: 'Dole',
    isActive: true
  }
*/

```

**Comment by ablankenship10** [comment]
Not sure if the nameOrOptions thing will work that way with typescript, but general idea being able to pass the name of a serializer thats already been configured or passing the schema directly on the fly if you wish.

**Comment by ablankenship10** [comment]
Separate include + exclude arrays may not be necessary. In the attached repository used for reference, they have some catch all values for things like 'All foreign keys' and the like, which may include fields you don't want that you then can add to the exclude array. May not be necessary for this library.

**Comment by fishcharlie** [comment]
@ablankenship10 Wow!! This looks amazing. Thank you so much for spending the time writing a detailed issue. Extremely easy to see what is going on and what the proposal is here.

Huge fan of this tho. I can already think of use cases where this would be **EXTREMELY** useful in my own projects.

I'm not sure if this makes sense as a plugin or as part of the core library. Currently, we don't have plugin support tho. We did in v1, but I gotta think about how we should handle it in v2 since I was not a fan of how it was done in v1. So for now, I think this can be part of the core library, and if we decide to do plugin support in the future, this might be up for consideration to spin off into a plugin.

> Model.addSerializer(name, options)

How do we feel about changing this to `Model.serializer.add`. We could also add an option for `Model.serializer.delete`. I can't think of a use case for it off the top of my head, but it'd give flexibility and would more closely align with the existing `Model.methods` system.

> Model.serialize(documents: Array, nameOrOptions: String/Object/Array)

Output of this will be an array correct?

> Document.serialize(nameOrOptions: String/Object/Array)

For this one. I'm assuming the output would be an object, correct? Not a Document instance?

> Not sure if the nameOrOptions thing will work that way with typescript

I think this should be fine. We already do this in a lot of places. Still figuring out all the ins and outs of TypeScript, but 99% sure this should be fine.

---

Couple other questions:

1. Would we want to integrate this deeper into Dynamoose? For example. Should you be allowed to call `await Model.get("ID", {"serialize": "mySerializer"})` and have it output a serialized object? Or should it purposefully be two different steps? I see pros and cons to both ways.
2. Looks like there are more options on the sequelize-to-json library. Has any thought gone into those?
3. It seems like when setting serialize options it should be `options.include` OR `options.exclude`. Or at least that would be the best practice that we would encourage and use in the documentation. Am I missing something there or?

**Comment by ablankenship10** [comment]
@fishcharlie Thanks thanks! Builtin or plugin would be fine by me either way, I do see it as something that should be possible without a plugin, its always been frustrating to clean up objects from different ORMs before. 

I like the idea of adding the methods under Model.serializer (and keeping document.serialize())

>> Model.serialize(documents: Array, nameOrOptions: String/Object/Array)

>Output of this will be an array correct?

Correct, it would be an array of the serialized objects (plain objects, no longer documents). Document.serialize() also returns  a plain object, no longer a document.

I forgot to mention a concept of a default serializer, so something like Model.serializer.default(name) could set the name of a default serializer to use, and automatically setting the default to the first one created. This way you could call document.serialize() without a name in most places. The only reason you would need different named schemas is for example outputting a users own user object with some sensitive information that only they should see versus outputting that same user object to another user searching for friends who doesn't necessarily need more sensitive info or the other users account settings fields. So you could use a different named schema in that case.

Other question answers:

1. I think there are cases where straight outputting the serialized data could be useful but could come at a later date. I believe its already possible to specify which fields you want returned directly from the database so thats kinda a way to control that already. The main point of the serializer IMO is that you need to request the entire object from the database, perform some logic using sensitive data such as authenticating the user with their password, and then output a cleaned version in the API response. 

2. The other main feature that library has is being able to deep serialize joined documents (the assoc option), so if your user had a field user.contacts where contacts was a list from another table, you could also define rules on how to serialize that document at the same time. I'm not familiar enough with Dynamo yet if that case is relevant or if they have any kind of join support, or if you've built in sorta pseudo-joins here. Basically we could just filter through all the fields and if any of them are other document instances, just also call serialize on them.

3. See my second comment above on that, and checkout Attribute Lists here: https://github.com/hauru/sequelize-to-json#attribute-lists. Basically they have these catch all attributes you can put in to either include or exclude several fields at once. The case where you would need to use both include and exclude at the same time is if you were to use one of those catch alls that included a group of fields but you also want to exclude one of them in the group.

**Comment by fishcharlie** [comment]
This all sounds great! Not sure when I'll have time to get around to implementing this. If anyone wants to submit a PR, that'd be awesome!! 😃

**Comment by ablankenship10** [comment]
I'll definitely try to pitch in this weekend. Are we at a good point in the typescript transition where it would make sense to add this feature? Wouldn't want to cause a complicated merge conflict here. Let me know what I should branch off of and I'll dive in. Thanks!

**Comment by ablankenship10** [comment]
How do you feel about Model.serializer being its own class instance rather than methods directly on the Model object?

**Comment by ablankenship10** [comment]
I developed the serialization class which should be mostly plug and play into the library, details in the gist description: https://gist.github.com/ablankenship10/152e33038fb29a0d452849e6a08758dd

It should fit your suggested changes to the API, and also I changed the options.postSerialize function to options.modify.

I'm not sure if my error handling is 100% thorough or if you'd even want it to throw an error or fail nicely. I'll keep away from working in the project for now while you're working on typescript stuff if you'd like to fit this in while you're doing that.
